Add winter:test command

RomainMazB opened the pull request:
First of the console command discussed here https://github.com/wintercms/winter/discussions/201

Accepts any review from anyone out there.

I did not found how it would be possible to display the phpunit running test. As of today it runs the tests "silently" and then output the results.

https://user-images.githubusercontent.com/53976837/121252668-93859c00-c8a8-11eb-961e-869d3c1f135e.mp4

RomainMazB committed:
Add plugin:test command

RomainMazB committed:
Remove author name

bennothommo commented:
@RomainMazB it's definitely possible to get the output happening in real-time - see the Dusk command class here, which does the same thing for Dusk tests: https://github.com/wintercms/wn-dusk-plugin/blob/main/console/Dusk.php. I would prefer that the output be in real-time if possible.

The key is to use the `Process` component from Symfony.

RomainMazB commented:
@bennothommo sure, I'd definitely prefer too. Thanks for pointed out how to do this, I'll definitely implement this

LukeTowers commented:
Might be nice to have it as `winter:test ?target` where target could be blank for the application's test suite or a specific plugin, and also forward any arguments / options / filters to the phpunit command (example: I use the `--filter` option a lot to run a specific single test)

RomainMazB committed:
Merge branch 'wintercms:develop' into plugintest

RomainMazB committed:
rename to winter:test and support core tests

RomainMazB committed:
Use Process to run live-tests

RomainMazB committed:
Proxy all options to phpunit

RomainMazB commented:
Ok, improvements made.
@LukeTowers to proxy all the phpunit filter we are forced to ignore validation (or create a 50 lines signature to describe all the available options). The same is done into the Laravel's core test command: https://github.com/nunomaduro/collision/blob/stable/src/Adapters/Laravel/Commands/TestCommand.php#L48-L53

A quick view of the final results.
https://user-images.githubusercontent.com/53976837/124030370-68e5b980-d9f6-11eb-893f-c6d57cc45b58.mp4

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+    {
+        parent::__construct();
+
+        // Because we need to proxy some options that are not defined
+        // We need to ignore validation errors
+        // Same is done in the native Laravel test commands
+        // @link https://github.com/nunomaduro/collision/blob/stable/src/Adapters/Laravel/Commands/TestCommand.php
+        $this->ignoreValidationErrors();
```
```suggestion
        /**
         * Ignore validation errors as option proxying is used by this command
         * @see https://github.com/nunomaduro/collision/blob/stable/src/Adapters/Laravel/Commands/TestCommand.php
         */
        $this->ignoreValidationErrors();
```

RomainMazB committed:
Change comment for ignoring validation

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+        // If the plugin argument is set, search for it and run its tests
+        if ($pluginArgument = $this->argument('plugin')) {
+            $pluginManager = PluginManager::instance();
+            $pluginName = $pluginManager->normalizeIdentifier($pluginArgument);
```
@RomainMazB the `exists()` method will already normalise the given plugin name, so no need to call this.
```suggestion
```

RomainMazB replied:
I'm not sure to understand this one. `exists()` method returns a boolean, not the plugin name 🤔

bennothommo replied:
@RomainMazB what I mean is that you do not need to normalize the plugin name before calling `exists` below - the `exists` method already normalizes the plugin name.

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+    public function handle()
+    {
+        // If the plugin argument is set, search for it and run its tests
+        if ($pluginArgument = $this->argument('plugin')) {
```
@RomainMazB where possible, I would like us to avoid assigning variables in a condition. It trips up Xdebug sometimes, and makes it difficult to breakpoint.
```suggestion
        $pluginName = $this->argument('plugin');
        if (!is_null($pluginName)) {
```

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+            $pluginName = $pluginManager->normalizeIdentifier($pluginArgument);
+
+            if (!$pluginManager->exists($pluginName)) {
+                throw new \InvalidArgumentException(sprintf('Plugin "%s" not found.', $pluginName));
```
@RomainMazB would it be better to just simply output an error as opposed to throwing an exception? A missing plugin is a fixable error.
```suggestion
                $this->error(sprintf('Plugin "%s" not found.', $pluginName));
```

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+
+            $pluginDir = $pluginManager->getPluginPath($pluginName);
+
+            $pluginPHPUnitXMLFile = $pluginDir . '/phpunit.xml';
+            if (!file_exists($pluginPHPUnitXMLFile)) {
+                throw new \InvalidArgumentException(sprintf('phpunit.xml file not found in "%s".', $pluginDir));
+            }
```
@RomainMazB same comment about outputting an error as above.

Also, note that distributed packages (like Winter plugins), can also have a `phpunit.xml.dist` file as well, allowing someone to locally use their own `phpunit.xml` if required, so it might be worth checking for the existence of this too.

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+             * Test plugin
+             */
+            $this->exec[] = "--configuration=$pluginPHPUnitXMLFile";
+            $this->output->writeln(sprintf('<info>Running  plugin\'s tests: %s.</info>', $pluginName));
```
```suggestion
            $this->info(sprintf('Running  plugin\'s tests: %s.', $pluginName));
```

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+            $this->output->writeln(sprintf('<info>Running  plugin\'s tests: %s.</info>', $pluginName));
+        }
+
+        // Else run winter's tests
```
```suggestion
        // Run the core Winter CMS tests
```

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+
+        // Else run winter's tests
+        else {
+            $this->output->writeln('<info>Running  Winter\'s core tests.</info>');
```
```suggestion
            $this->info('Running  Winter\'s core tests.');
```

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+use Symfony\Component\Console\Input\InputArgument;
+
+/**
+ * Console command to test a plugin or winter's core.
```
```suggestion
 * Console command to test a plugin or the Winter CMS core.
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+
+            $pluginDir = $pluginManager->getPluginPath($pluginName);
+
+            $pluginPHPUnitXMLFile = $pluginDir . '/phpunit.xml';
```
Might also have to look for the phpunit.xml.dist or whatever it is that we're using on the winter storm library

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+/**
+ * Console command to test a plugin or winter's core.
+ *
+ * This searches for a phpunit.xml file into the plugin's directory and run its tests
```
```suggestion
 * If a plugin is provided, this command will search for a `phpunit.xml` file inside the plugin's directory and run its tests.
```

bennothommo commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,112 @@
+     * The console command description.
+     * @var string
+     */
+    protected $description = "Run tests of Winter's core or an existing plugin.";
```
```suggestion
    protected $description = "Run tests for the Winter CMS core or an existing plugin.";
```

LukeTowers commented:
@RomainMazB does the core tests differentiate between the CMS tests and the library tests? @bennothommo should we have a way to trigger them separately or together?

bennothommo commented:
@LukeTowers we could define all the tests as separate suites in the `phpunit.xml` file and it would allow us to run all tests together, but there's no way to separate the test results - not sure if that's necessary a problem for people.

We used to combine the library and core tests, but I separated them during the L6 update because tests were failing when core PRs were relying on unmerged library PRs.

RomainMazB committed:
Improve comments

Co-authored-by: Ben Thomson <ben@abweb.com.au>

RomainMazB committed:
Improve command description

Co-authored-by: Ben Thomson <ben@abweb.com.au>

RomainMazB committed:
Improve class comment

Co-authored-by: Ben Thomson <ben@abweb.com.au>

RomainMazB committed:
Improve comments

Co-authored-by: Ben Thomson <ben@abweb.com.au>

RomainMazB committed:
Change output info with info shortcut

Co-authored-by: Ben Thomson <ben@abweb.com.au>

RomainMazB committed:
Apply suggestions from code review

Co-authored-by: Ben Thomson <ben@abweb.com.au>

RomainMazB committed:
Gives the ability to pass a custom configuration file, and now fallback to phpunit.xml.dist file

RomainMazB commented:
In the latest commit, we can now pass a custom configuration file using `--configuration` option.

If none is defined, it looks for `phpunit.xml` file and fallback to `phpunit.xml.dist`.

Need precision about my comment on @bennothommo review about `exists()` and the final thoughts on the library and core tests to continue.

RomainMazB committed:
Fix indentation

RomainMazB committed:
Remove useless part of a comment

jaxwilko committed:
Reworked command argument handling and test execution

jaxwilko commented:
I've made a few changes to improve the arguments passthrough to phpunit, there is an issue where if phpunit arguments are passed before the command arguments then the command args aren't parsed.

Seems to be an issue from `Symfony\Component\Console\Input\ArgvInput`. I shall look into it and see what I can find :)

RomainMazB commented:
@LukeTowers @bennothommo.
I've been toying with the `ArgvInput` symfony component to find the cause of the issue mentioned by @jaxwilko.

The issue is not directly the fact that we pass a PHPUnit argument before `--plugin` or `--core`, but the fact the `winter:test` command doesn't define those options, which raise a `RuntimeException`, the exception is silently hidden by the `ignoreValidationErrors` rule we are using and present in the Symfony Command class here: https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/Console/Command/Command.php#L248-L252

This makes the options to not be correctly set and runs the command without any options at all. I'm afraid that we won't be able to allow phpunit options before the `winter:test` command's ones.

I've added a section in the documentation emphasizing the fact that [phpunit options should be passed **after** the command's options.](https://github.com/wintercms/docs/pull/34/files#diff-8fe50b3d6c3c6248eeb1df04812987226c7bafc5f5305b48277e5ee6f79eaf7aR335-R337) which works well:
![image](https://user-images.githubusercontent.com/53976837/137711014-e32e5d5d-cc09-44ff-befe-107deef32616.png)

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+
+        if ($this->option('core')) {
+            if (!$configs['core']) {
+                throw new ApplicationException('unable to find core `phpunit.xml`. Try downloading it from github.');
```
```suggestion
                throw new ApplicationException("Unable to find the core's phpunit.xml file. Try downloading it from GitHub.");
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+
+        if ($plugin = $this->option('plugin')) {
+            if (!isset($configs['plugins'][strtolower($plugin)])) {
+                throw new ApplicationException(sprintf('unable to find %s\'s `phpunit.xml`', $plugin));
```
```suggestion
                throw new ApplicationException(sprintf("Unable to find %s\'s phpunit.xml file', $plugin));
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+                continue;
+            }
+
+            $this->info('Running tests for: Winter CMS ' . $type);
```
```suggestion
            $this->info('Running tests for Winter CMS: ' . $type);
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+    protected function getOptions(): array
+    {
+        return [
+            ['plugin', 'p', InputOption::VALUE_OPTIONAL, 'The name of the plugin. Eg: AuthorName.PluginName', null],
```
```suggestion
            ['plugin', 'p', InputOption::VALUE_OPTIONAL, 'The name of the plugin. Ex: AuthorName.PluginName', null],
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+    {
+        return [
+            ['plugin', 'p', InputOption::VALUE_OPTIONAL, 'The name of the plugin. Eg: AuthorName.PluginName', null],
+            ['configuration', 'c', InputOption::VALUE_OPTIONAL, 'The path to a phpunit xml config', null],
```
```suggestion
            ['configuration', 'c', InputOption::VALUE_OPTIONAL, 'The path to a PHPUnit XML config file', null],
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+        return [
+            ['plugin', 'p', InputOption::VALUE_OPTIONAL, 'The name of the plugin. Eg: AuthorName.PluginName', null],
+            ['configuration', 'c', InputOption::VALUE_OPTIONAL, 'The path to a phpunit xml config', null],
+            ['core', 'o', InputOption::VALUE_NONE, 'Test the core'],
```
```suggestion
            ['core', 'o', InputOption::VALUE_NONE, 'Run the Winter CMS core tests'],
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     */
+    protected function execPhpUnit(string $config, array $args): int
+    {
+        // find and bind the phpunit executable
```
```suggestion
        // Find and bind the phpunit executable
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+            null
+        );
+
+        // attempt to set tty mode, if unsupported catch and warn the exception message
```
```suggestion
        // Attempt to set tty mode, catch and warn with the exception message if unsupported
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * Path to phpunit binary
+     * @var ?string
+     */
+    protected $phpUnitExec = null;
```
```suggestion
    /**
     * The console command name.
     * @var string
     */
    protected $name = 'winter:test';

    /**
     * @var string The console command signature as ignoreValidationErrors causes options not to be registered.
     */
    protected $signature = 'winter:test {?--p|plugin=} {?--c|configuration=} {?--o|core}';

    /**
     * @var string The console command description.
     */
    protected $description = 'Run tests for the Winter CMS core or an existing plugin.';

    /**
     * @var ?string Path to phpunit binary
     */
    protected $phpUnitExec = null;
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * @throws ApplicationException
+     *
+     * @return int|void
+     */
```
```suggestion
    /**
     * Execute the console command.
     *
     * @throws ApplicationException
     * @return int|void
     */
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * @param array $args array of params for phpunit
+     *
+     * @return int exit code from process
+     */
```
```suggestion
    /**
     * Execute a phpunit test
     *
     * @param string $config Path to configuration file
     * @param array $args Array of params for PHPUnit
     * @return int Exit code from process
     */
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * @param string $path
+     *
+     * @return ?string
+     */
```
```suggestion
    /**
     * Search for the config file to use.
     * Priority order is: phpunit.xml, phpunit.xml.dist
     */
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+    /**
+     * Strips out commands arguments and options in order to return arguments/options for PHPUnit.
+     *
+     * @return array
```
```suggestion
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * @param array $arguments
+     * @param string $remove
+     *
+     * @return array
```
```suggestion
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * Find all PHPUnit config files (core, lib, plugins)
+     *
+     * @return array
+     */
```
```suggestion
    /**
     * Find all PHPUnit config files (core, lib, plugins)
     */
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * Get the console command arguments.
+     *
+     * @return array
+     */
```
```suggestion
     * Get the console command arguments.
     */
```

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,282 @@
+     * Get the console command options.
+     *
+     * @return array
+     */
```
```suggestion
    /**
     * Get the console command options.
     */
```

LukeTowers committed:
Apply suggestions from code review

No need to include docblock comments for information provided by type hints unless additional comments are made that justify a docblock comment

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,263 @@
+
+        if ($plugin = $this->option('plugin')) {
+            if (!isset($configs['plugins'][strtolower($plugin)])) {
+                throw new ApplicationException(sprintf("Unable to find %s\'s phpunit.xml file', $plugin));
```
```suggestion
                throw new ApplicationException(sprintf("Unable to find %s\'s phpunit.xml file", $plugin));
```

LukeTowers committed:
Apply suggestions from code review

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,263 @@
+    {
+        return [];
+    }
+
```
```suggestion
```

LukeTowers committed:
Update modules/system/console/WinterTest.php

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,255 @@
+use Winter\Storm\Exception\ApplicationException;
+
+/**
+ * Console command to test a plugin or the Winter CMS core.
```
```suggestion
 * Console command to run tests for plugins or the Winter CMS core.
```

LukeTowers committed:
Update modules/system/console/WinterTest.php

LukeTowers commented on modules/system/console/WinterTest.php:
```diff
@@ -0,0 +1,255 @@
+    /**
+     * The console command name.
+     * @var string
+     */
```
```suggestion
    /**
     * @var string The console command name.
     */
```

LukeTowers committed:
Update modules/system/console/WinterTest.php
